### PR TITLE
Fix ComponentStyle caching strategy to take StyleSheet cache into account

### LINKED ISSUE #1634: Style Cache not invalidated when StyleSheetManager target is removed
## System:
 - OS: macOS High Sierra 10.13.3
 - CPU: x64 Intel(R) Core(TM) i7-2860QM CPU @ 2.50GHz
 - Free Memory: 494.87 MB
 - Total Memory: 8.00 GB
 - Shell: /usr/local/bin/fish - 2.7.1
## Binaries:
 - Node: 8.9.4
 - Yarn: 1.5.1
 - npm: 5.7.1
 - Watchman: 4.7.0
### styled-components:
 - wanted: ^3.2.3
 - installed: 3.2.3
## Reproduction

I'm using styled-components within an iframe on a PowerBI visual. These are instantiated pretty frequently as you switch around report pages. I'm using the StyleSheetManager to target my iframe and inject styles into the head. This works the first time, but when this frame re-renders the styles are gone from the iframe document, but styled-components is still holding onto the cached names.

## Steps to reproduce
Working on a test to reproduce this

## Expected Behavior
Styled-components should inject styles into newly created iframes if the old one disappears.

Perhaps ComponentStyle checks that the StyleSheetManager instance is the same instance it registered its styles with before?

## Actual Behavior
styled-components has cached stylenames, but those style names are no longer available.

### PR DESCRIPTION
**Pull request by darthtrevino**
Fixes #1634, Whitespace-change free version of PR #1635

When ComponentStyle.generateAndInjectStyles is invoked, the stylesheet instance should be checked even if the style name is already defined. If the component is being rendered in a sibling iframe with a separate StyleSheetManager, this allows the second iframe to receive the injected styles.

### CODE REVIEW

**Comment by kitten** [comment]
@darthtrevino Changing the StyleSheet while the component is mounted — essentially hot swapping the sheet — is not something we should support, just to be clear again. However, this is a reasonably small change that I can maybe accept, but we'll have to regression test it before release. Hence a small disclaimer: Sorry in advance if this change doesn't land :wink:

The current state of the code breaks an important caching optimisation, as it runs a bunch of operations that we're trying to avoid just to fix this. I'll leave some comments on the code itself since there's another way.

File: src/models/test/StyleSheetManager.test.js [hunk 1]
```diff
@@ -116,6 +119,69 @@ describe('StyleSheetManager', () => {
     expect(document.body.innerHTML).toMatchSnapshot()
   })
 
+  // https://github.com/styled-components/styled-components/issues/1634
+  it('should flush styles to an iframe when the iframe is re-rendered', async () => {
```
Comment (kitten) [inline src/models/test/StyleSheetManager.test.js]: This test is extremely specific to your environment and additionally this conflicts with it as well:

> Changing the StyleSheet while the component is mounted — essentially hot swapping the sheet — is not something we should support,

Also it's a little unnecessarily complex. Instead we can use enzyme to swap out the the stylesheet itself together with a prop, and expect to see a different class.

Edit: I'd say it's ok to not test this for now. I can write a test eventually, but one that is specific to all caching guarantees instead

File: src/models/ComponentStyle.js [hunk 2]
```diff
@@ -78,12 +78,14 @@ export default (
      * */
     generateAndInjectStyles(executionContext: Object, styleSheet: StyleSheet) {
       const { isStatic, componentId, lastClassName } = this
-      if (areStylesCacheable && isStatic && lastClassName !== undefined) {
-        return lastClassName
-      }
-
```
Comment (kitten) [inline src/models/ComponentStyle.js]: Here you're leading the code into expensive operations that we're trying to avoid

File: src/models/ComponentStyle.js [hunk 3]
```diff
@@ -78,12 +78,14 @@ export default (
...
       const flatCSS = flatten(this.rules, executionContext)
-      const name = generateRuleHash(this.componentId + flatCSS.join(''))
+
+      const useLastClassName: boolean =
```
Comment (kitten) [inline src/models/ComponentStyle.js]: This goes through all operations even if it's not necessary too

**Comment by darthtrevino** [comment]
I don't believe this will hot-swap stylesheets for mounted and styled components. This is kind of analogous to a Java thread-local - it's two components of the same type in totally different stylesheet contexts.

File: src/models/ComponentStyle.js [hunk 4]
```diff
@@ -78,12 +78,14 @@ export default (
      * */
     generateAndInjectStyles(executionContext: Object, styleSheet: StyleSheet) {
       const { isStatic, componentId, lastClassName } = this
-      if (areStylesCacheable && isStatic && lastClassName !== undefined) {
```
Comment (kitten) [inline src/models/ComponentStyle.js]: This is the crux of it, so instead you can replace this: 

`if (areStylesCacheable && isStatic && lastClassName !== undefined) {`

with this:

`if (areStylesCacheable && isStatic && lastClassName !== undefined && styleSheet.hasNameForId(componentId, lastClassName)) {`

a lot simpler, doesn't break our caching and optimisation, and it should have the same effect :+1:

so essentially this is the only change that is needed

**Comment by darthtrevino** [comment]
I'm guessing the flatten() is what you're trying to avoid?

**Comment by kitten** [comment]
@darthtrevino 

> I don't believe this will hot-swap stylesheets for mounted and styled components. This is kind of analogous to a Java thread-local - it's two components of the same type in totally different stylesheet contexts.

No, this code change won't. Essentially you're just invalidating the cache for a new stylesheet. Whether that's hot swapped through the context, or because you're suddenly in a new context with the same ComponentStyle doesn't matter that much, because it won't occur during normal usage.

I think it's a fair addition though, so I've left some comments above and I think we can ship this

**Comment by darthtrevino** [comment]
No problem, I get that it's a weird use case. I put some additional context at the tail of the last PR

**Comment by kitten** [comment]
@darthtrevino just seen your last changes. Haven’t given it a quick run yet but see you sure that the change I proposed here won’t be enough? https://github.com/styled-components/styled-components/pull/1636#comment-176869902

Because essentially we only have to prevent returning the lastClassName if it’s not on the StyleSheet so that one line change should be enough

**Comment by darthtrevino** [comment]
@kitten Switching to your suggestion - I think that will be fine

File: CHANGELOG.md [hunk 5]
```diff
@@ -6,7 +6,7 @@ All notable changes to this project will be documented in this file. If a contri
 
 ## Unreleased
 
-- n/a
+- Fix issue with multiple iframes and StyleSheetManager instances. @darthtrevino (see [#1634])
```
Comment (kitten) [inline CHANGELOG.md]: Can we make this generic as well like the test?

I.e. fix last className caching strategy to take StyleSheet cache into account

Or sth

**Comment by kitten** [comment]
@darthtrevino thanks for making this change this quickly 👏 hope it will help more ppl with similar edge cases.

I’ll give it a couple of tests tomorrow and will push a prerelease. If that looks fine I’ll publish a patch :)

**Comment by bliitzkrieg** [comment]
This fix should resolve a similar issue I am having with a different use case. I created an embedded component system that mounts a react component (bundled with styled components) on html markup on any given client. The components wrap a StyleSheetManager that use the element its mounting as the target and when I programically unmount/mount the component for advanced usage, the styles disappear.

**Comment by kitten** [comment]
@bliitzkrieg yes, that might be fixed as well by this. Quick note that what you're doing will break specificity and certain guarantees (see previous comments) :wink:

**Comment by mxstbr** [comment]
Thank you so much for helping us improve styled-components! Based on our [Community Guidelines](https://github.com/styled-components/styled-components/blob/master/CONTRIBUTING.md) every person that has a PR of any kind merged is offered **an invitation to the styled-components organization**—that is you right now!

Should you accept, you'll get write access to the main repository. (and a fancy `styled-components` logo on your GitHub profile!) You'll be able to label and close issues, merge pull requests, create new branches, etc. We want you to help us out with those things, so don't be scared to do them! Make sure to read our [contribution and community guidelines](https://github.com/styled-components/styled-components/blob/master/CONTRIBUTING.md), which explains all of this in more detail. Of course, if you have any questions just let me know!
